File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-gitee/src/main/java/com/tencent/devops/scm/provider/git/gitee/GiteeWebhookEnricher.java
package com.tencent.devops.scm.provider.git.gitee;

import com.tencent.devops.scm.api.WebhookEnricher;
import com.tencent.devops.scm.api.exception.UnAuthorizedScmApiException;
import com.tencent.devops.scm.api.function.TriConsumer;
import com.tencent.devops.scm.api.pojo.Change;
import com.tencent.devops.scm.api.pojo.PullRequest;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository;
import com.tencent.devops.scm.api.pojo.webhook.Webhook;
import com.tencent.devops.scm.api.pojo.webhook.git.GitPushHook;
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestHook;
import com.tencent.devops.scm.provider.git.gitee.auth.GiteeTokenAuthProviderAdapter;
import com.tencent.devops.scm.sdk.gitee.GiteeApi;
import com.tencent.devops.scm.sdk.gitee.GiteeApiException;
import com.tencent.devops.scm.sdk.gitee.GiteeApiFactory;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteeCommitCompare;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteePullRequest;
import com.tencent.devops.scm.sdk.gitee.pojo.GiteePullRequestDiff;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;

public class GiteeWebhookEnricher implements WebhookEnricher {
    private final Map<Class<? extends Webhook>, TriConsumer<GiteeApi, GitScmProviderRepository, Webhook>> eventActions;

    private final GiteeApiFactory apiFactory;

    public GiteeWebhookEnricher(GiteeApiFactory apiFactory) {
        this.apiFactory = apiFactory;
        this.eventActions = new HashMap<>();
        eventActions.put(PullRequestHook.class, this::enrichPullRequestHook);
        eventActions.put(GitPushHook.class, this::enrichPushHook);
    }

    private void enrichPushHook(GiteeApi giteeApi, GitScmProviderRepository repository, Webhook hook) {
        GitPushHook gitPushHook = (GitPushHook) hook;
        GiteeCommitCompare commitCompare = giteeApi.getFileApi().commitCompare(
                repository.getProjectIdOrPath(),
                gitPushHook.getBefore(),
                gitPushHook.getAfter(),
                false
        );
        gitPushHook.setChanges(GiteeObjectConverter.convertCompare(commitCompare));
        gitPushHook.setChanges(GiteeObjectConverter.convertCompare(commitCompare));
    }

    private void enrichPullRequestHook(GiteeApi giteeApi, GitScmProviderRepository repository, Webhook hook) {
        PullRequestHook pullRequestHook = (PullRequestHook) hook;
        PullRequest pullRequest = pullRequestHook.getPullRequest();
        GiteePullRequestDiff[] pullRequestDiffs = giteeApi.getPullRequestApi()
                .listChanges(repository.getProjectIdOrPath(), pullRequest.getNumber().longValue());
        List<Change> changes = Arrays.stream(pullRequestDiffs)
                .map(GiteeObjectConverter::convertChange)
                .collect(Collectors.toList());
        pullRequestHook.setChanges(changes);
        // 根据完整的MR/Review信息填充变量
        pullRequestHook.getExtras().putAll(
                fillPullRequestReviewVars(
                        giteeApi,
                        repository,
                        pullRequest
                )
        );
    }

    @Override
    public Webhook enrich(ScmProviderRepository repository, Webhook webhook) throws UnAuthorizedScmApiException {
        GitScmProviderRepository repo = (GitScmProviderRepository) repository;
        try {
            GiteeApi giteeApi = apiFactory.fromAuthProvider(GiteeTokenAuthProviderAdapter.get(repo.getAuth()));
            if (eventActions.containsKey(webhook.getClass())) {
                eventActions.get(webhook.getClass()).accept(giteeApi, repo, webhook);
            }
            return webhook;
        } catch (GiteeApiException exception) {
            if (exception.getStatusCode() == 401 || exception.getStatusCode() == 403) {
                throw new UnAuthorizedScmApiException(exception.getMessage());
            } else {
                throw exception;
            }
        }
    }

    public Map<String,Object> fillPullRequestReviewVars(
            GiteeApi giteeApi,
            GitScmProviderRepository repository,
            PullRequest pullRequest
    ) {
        return fillPullRequestVars(giteeApi, repository, pullRequest);
    }

    public Map<String,Object> fillPullRequestVars(
            GiteeApi giteeApi,
            GitScmProviderRepository repository,
            PullRequest pullRequest
    ) {
        GiteePullRequest giteePullRequest = giteeApi.getPullRequestApi().getPullRequest(
                repository.getProjectIdOrPath(),
                pullRequest.getNumber().longValue()
        );
        Map<String, Object> outputs = new HashMap<>();
        if (giteePullRequest != null) {
            pullRequest.setTitle(StringUtils.defaultString(giteePullRequest.getTitle(), ""));
            outputs.putAll(GiteeObjectToMapConverter.convertPullRequest(giteePullRequest));
        }
        return outputs;
    }
}

File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-gitee/src/main/kotlin/com/tencent/devops/scm/provider/git/gitee/GiteeWebhookEnricher.kt
package com.tencent.devops.scm.provider.git.gitee

import com.tencent.devops.scm.api.WebhookEnricher
import com.tencent.devops.scm.api.enums.EventAction
import com.tencent.devops.scm.api.pojo.PullRequest
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.api.pojo.repository.git.GitScmProviderRepository
import com.tencent.devops.scm.api.pojo.webhook.Webhook
import com.tencent.devops.scm.api.pojo.webhook.git.GitPushHook
import com.tencent.devops.scm.api.pojo.webhook.git.PullRequestHook
import com.tencent.devops.scm.sdk.gitee.GiteeApi
import com.tencent.devops.scm.sdk.gitee.GiteeApiFactory

class GiteeWebhookEnricher(private val apiFactory: GiteeApiFactory) : WebhookEnricher {

    private val eventActions = mutableMapOf<Class<out Webhook>, (GitScmProviderRepository, Webhook) -> Unit>()

    init {
        eventActions[PullRequestHook::class.java] = this::enrichPullRequestHook
        eventActions[GitPushHook::class.java] = this::enrichPushHook
    }

    private fun enrichPushHook(repository: GitScmProviderRepository, hook: Webhook) {
        val gitPushHook = hook as GitPushHook
        if (gitPushHook.action == EventAction.PUSH_FILE) {
            GiteeApiTemplate.execute(repository, apiFactory) { _, giteeApi ->
                val commitCompare = giteeApi.fileApi.commitCompare(
                    repository.projectIdOrPath,
                    gitPushHook.before,
                    gitPushHook.after,
                    false
                )
                gitPushHook.changes = GiteeObjectConverter.convertCompare(commitCompare)
            }
        }
    }

    private fun enrichPullRequestHook(repository: GitScmProviderRepository, hook: Webhook) {
        val pullRequestHook = hook as PullRequestHook
        val pullRequest = pullRequestHook.pullRequest
        GiteeApiTemplate.execute(repository, apiFactory) { _, giteeApi ->
            // 填充文件变更列表
            val pullRequestDiffs = giteeApi.pullRequestApi.listChanges(
                repository.projectIdOrPath,
                pullRequest.number.toLong()
            )
            pullRequestHook.changes = pullRequestDiffs.map { GiteeObjectConverter.convertChange(it) }

            // 填充最后提交信息
            val requestHookCommit = pullRequestHook.commit
            val commit = giteeApi.commitApi.getCommit(
                repository.projectIdOrPath,
                requestHookCommit.sha
            )
            requestHookCommit.message = commit.commit?.message ?: ""
            requestHookCommit.author = GiteeObjectConverter.convertSignature(commit.author)

            // 根据完整的MR/Review信息填充变量
            pullRequestHook.extras.putAll(
                fillPullRequestReviewVars(giteeApi, repository, pullRequest)
            )
        }
    }

    override fun enrich(repository: ScmProviderRepository, webhook: Webhook): Webhook {
        eventActions[webhook.javaClass]?.invoke(repository as GitScmProviderRepository, webhook)
        return webhook
    }

    private fun fillPullRequestReviewVars(
        giteeApi: GiteeApi,
        repository: GitScmProviderRepository,
        pullRequest: PullRequest
    ): Map<String, Any> {
        return fillPullRequestVars(giteeApi, repository, pullRequest)
    }

    private fun fillPullRequestVars(
        giteeApi: GiteeApi,
        repository: GitScmProviderRepository,
        pullRequest: PullRequest
    ): Map<String, Any> {
        return giteeApi.pullRequestApi.getPullRequest(
            repository.projectIdOrPath,
            pullRequest.number.toLong()
        )?.let {
            pullRequest.title = it.title ?: ""
            GiteeObjectToMapConverter.convertPullRequest(it)
        } ?: mapOf()
    }
}
